File: app/src/main/java/dev/ragnarok/fenrir/api/interfaces/IUsersApi.java
package dev.ragnarok.fenrir.api.interfaces;

import androidx.annotation.CheckResult;

import java.util.Collection;
import java.util.List;

import dev.ragnarok.fenrir.api.model.Items;
import dev.ragnarok.fenrir.api.model.VKApiGift;
import dev.ragnarok.fenrir.api.model.VKApiStory;
import dev.ragnarok.fenrir.api.model.VKApiUser;
import dev.ragnarok.fenrir.api.model.response.StoryResponse;
import dev.ragnarok.fenrir.api.model.server.VkApiStoryUploadServer;
import io.reactivex.rxjava3.core.Single;


public interface IUsersApi {

    @CheckResult
    Single<VKApiUser> getUserWallInfo(int userId, String fields, String nameCase);

    @CheckResult
    Single<Items<VKApiUser>> getFollowers(Integer userId, Integer offset, Integer count,
                                          String fields, String nameCase);

    @CheckResult
    Single<Items<VKApiUser>> getRequests(Integer offset, Integer count, Integer extended, Integer out, String fields);

    @CheckResult
    Single<Items<VKApiUser>> search(String query, Integer sort, Integer offset, Integer count,
                                    String fields, Integer city, Integer country, String hometown,
                                    Integer universityCountry, Integer university, Integer universityYear,
                                    Integer universityFaculty, Integer universityChair, Integer sex,
                                    Integer status, Integer ageFrom, Integer ageTo, Integer birthDay,
                                    Integer birthMonth, Integer birthYear, Boolean online,
                                    Boolean hasPhoto, Integer schoolCountry, Integer schoolCity,
                                    Integer schoolClass, Integer school, Integer schoolYear,
                                    String religion, String interests, String company,
                                    String position, Integer groupId, String fromList);

    @CheckResult
    Single<List<VKApiUser>> get(Collection<Integer> userIds, Collection<String> domains,
                                String fields, String nameCase);

    @CheckResult
    Single<VkApiStoryUploadServer> stories_getPhotoUploadServer();

    @CheckResult
    Single<VkApiStoryUploadServer> stories_getVideoUploadServer();

    @CheckResult
    Single<Items<VKApiStory>> stories_save(String upload_results);

    @CheckResult
    Single<StoryResponse> getStory(Integer owner_id, Integer extended, String fields);

    @CheckResult
    Single<Integer> checkAndAddFriend(Integer userId);

    @CheckResult
    Single<Items<VKApiGift>> getGifts(Integer user_id, Integer count, Integer offset);

    @CheckResult
    Single<StoryResponse> searchStory(String q, Integer mentioned_id, Integer count, Integer extended, String fields);

    @CheckResult
    Single<Integer> report(Integer userId, String type, String comment);

}

File: app/src/main/java/dev/ragnarok/fenrir/api/interfaces/IUsersApi.kt
package dev.ragnarok.fenrir.api.interfaces

import androidx.annotation.CheckResult
import dev.ragnarok.fenrir.api.model.Items
import dev.ragnarok.fenrir.api.model.VKApiGift
import dev.ragnarok.fenrir.api.model.VKApiStory
import dev.ragnarok.fenrir.api.model.VKApiUser
import dev.ragnarok.fenrir.api.model.response.StoryResponse
import dev.ragnarok.fenrir.api.model.server.VkApiStoryUploadServer
import io.reactivex.rxjava3.core.Single

interface IUsersApi {
    @CheckResult
    fun getUserWallInfo(userId: Int, fields: String?, nameCase: String?): Single<VKApiUser>

    @CheckResult
    fun getFollowers(
        userId: Int?, offset: Int?, count: Int?,
        fields: String?, nameCase: String?
    ): Single<Items<VKApiUser>>

    @CheckResult
    fun getRequests(
        offset: Int?,
        count: Int?,
        extended: Int?,
        out: Int?,
        fields: String?
    ): Single<Items<VKApiUser>>

    @CheckResult
    fun search(
        query: String?, sort: Int?, offset: Int?, count: Int?,
        fields: String?, city: Int?, country: Int?, hometown: String?,
        universityCountry: Int?, university: Int?, universityYear: Int?,
        universityFaculty: Int?, universityChair: Int?, sex: Int?,
        status: Int?, ageFrom: Int?, ageTo: Int?, birthDay: Int?,
        birthMonth: Int?, birthYear: Int?, online: Boolean?,
        hasPhoto: Boolean?, schoolCountry: Int?, schoolCity: Int?,
        schoolClass: Int?, school: Int?, schoolYear: Int?,
        religion: String?, interests: String?, company: String?,
        position: String?, groupId: Int?, fromList: String?
    ): Single<Items<VKApiUser>>

    @CheckResult
    operator fun get(
        userIds: Collection<Int>?, domains: Collection<String>?,
        fields: String?, nameCase: String?
    ): Single<List<VKApiUser>>

    @CheckResult
    fun stories_getPhotoUploadServer(): Single<VkApiStoryUploadServer>

    @CheckResult
    fun stories_getVideoUploadServer(): Single<VkApiStoryUploadServer>

    @CheckResult
    fun stories_save(upload_results: String?): Single<Items<VKApiStory>>

    @CheckResult
    fun getStory(owner_id: Int?, extended: Int?, fields: String?): Single<StoryResponse>

    @CheckResult
    fun checkAndAddFriend(userId: Int?): Single<Int>

    @CheckResult
    fun getGifts(user_id: Int?, count: Int?, offset: Int?): Single<Items<VKApiGift>>

    @CheckResult
    fun searchStory(
        q: String?,
        mentioned_id: Int?,
        count: Int?,
        extended: Int?,
        fields: String?
    ): Single<StoryResponse>

    @CheckResult
    fun report(userId: Int?, type: String?, comment: String?): Single<Int>
}